lib/core/src/restore.rs

```diff
@@ -0,0 +1,1164 @@
+            Some(_) => match &self.claim_tx_id {
+                Some(_) => PaymentState::Complete,
+                None => match &self.refund_tx_id {
+                    Some(refund_tx_id) => match refund_tx_id.confirmed {
+                        true => PaymentState::Failed,
+                        false => PaymentState::RefundPending,
+                    },
+                    None => PaymentState::Pending,
+                },
+            },
+            None => PaymentState::TimedOut,
+        }
+    }
+}
+
+pub(crate) struct RecoveredOnchainDataReceive {
+    lockup_tx_id: Option<HistoryTxId>,
+    claim_tx_id: Option<HistoryTxId>,
+}
+impl PartialSwapState for RecoveredOnchainDataReceive {
+    fn get_partial_state(&self) -> PaymentState {
+        match (&self.lockup_tx_id, &self.claim_tx_id) {
+            (Some(_), Some(claim_tx_id)) => match claim_tx_id.confirmed {
+                true => PaymentState::Complete,
+                false => PaymentState::Pending,
+            },
+            (Some(_), None) => PaymentState::Pending,
+            // TODO How to distinguish between Failed and Created (if in both cases, no lockup or claim tx present)
+            //   See https://docs.boltz.exchange/v/api/lifecycle#reverse-submarine-swaps
+            _ => PaymentState::Failed,
+        }
+    }
+}
+
+pub(crate) struct RecoveredOnchainDataChainSend {
+    lbtc_user_lockup_tx_id: Option<HistoryTxId>,
+    lbtc_refund_tx_id: Option<HistoryTxId>,
+    btc_server_lockup_tx_id: Option<HistoryTxId>,
+    btc_claim_tx_id: Option<HistoryTxId>,
+}
+impl PartialSwapState for RecoveredOnchainDataChainSend {
+    fn get_partial_state(&self) -> PaymentState {
+        // TODO How to detect TimedOut state?
+        //     TimedOut: This covers the case when the swap state is still Created and the swap fails to reach the
+        //     Pending state in time. The TimedOut state indicates the lockup tx should never be broadcast.
+        match &self.lbtc_user_lockup_tx_id {
+            Some(_) => match &self.btc_claim_tx_id {
+                Some(_) => PaymentState::Complete,
+                None => match &self.lbtc_refund_tx_id {
+                    Some(tx) => match tx.confirmed {
+                        true => PaymentState::Failed,
+                        false => PaymentState::RefundPending,
+                    },
+                    None => PaymentState::Created,
+                },
+            },
+            None => PaymentState::Created,
+        }
+    }
+}
+
+pub(crate) struct RecoveredOnchainDataChainReceive {
+    lbtc_server_lockup_tx_id: Option<HistoryTxId>,
+    lbtc_server_claim_tx_id: Option<HistoryTxId>,
+    btc_user_lockup_tx_id: Option<HistoryTxId>,
+    btc_refund_tx_id: Option<HistoryTxId>,
+}
+impl PartialSwapState for RecoveredOnchainDataChainReceive {
+    fn get_partial_state(&self) -> PaymentState {
+        match &self.btc_user_lockup_tx_id {
+            Some(_) => match &self.lbtc_server_claim_tx_id {
```

Comment:
Do we care about the server claim? As far as I understand it is the last step in the swap which we don't care about. The swap completes when we claim If I get it right.

Reply:
After reading all the code I now see it is not the server claim but is our claim.
So perhaps rename to `lbtc_user_claim_tx_id` ?

Reply:
> After reading all the code I now see it is not the server claim but is our claim. So perhaps rename to `lbtc_user_claim_tx_id` ?

I checked with @dangeross and this is the server claim as per Boltz nomenclature. For them, the initiator (not creator) of the swap is the "user", and txs associated with those funds are user txs. In this case, that's the BTC lockup tx, BTC claim tx and even BTC refund tx.

Conversely, the funds locked by the swapper ("server") are associated with server txs. In this case, that's the LBTC lockup tx and LBTC claim tx (the tx you referred to).

Reply:
> Do we care about the server claim? As far as I understand it is the last step in the swap which we don't care about. The swap completes when we claim If I get it right.

The server claim is our claim, that's the only reason we check this.

Reply:
As discussed in the group chat, this is correct as it is, so marking this as resolved.